subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIsolationTest.groovy

```diff
@@ -0,0 +1,194 @@
+                    println "variants: " + first.collect { it.variant.attributes }
+                }
+            }
+            
+            task increment {
+                doFirst {
+                    // Just to show that incrementing the counter doesn't matter.
+                    counter.increment()
+                }
+            }
+
+            task resolveSecond(type: Copy) {
+                def second = configurations.compile.incoming.artifactView {
+                    attributes { it.attribute(artifactType, 'secondCount') }
+                }.artifacts
+                from second.artifactFiles
+                into "\${buildDir}/libs2"
+                
+                doLast {
+                    println "files: " + second.collect { it.file.name }
+                    println "ids: " + second.collect { it.id }
+                    println "components: " + second.collect { it.id.componentIdentifier }
+                    println "variants: " + second.collect { it.variant.attributes }
+                }
+            }
+
+            task resolveThird(type: Copy) {
+                def third = configurations.compile.incoming.artifactView {
+                    attributes { it.attribute(artifactType, 'thirdCount') }
+                }.artifacts
+                from third.artifactFiles
+                into "\${buildDir}/libs3"
+                
+                doLast {
+                    println "files: " + third.collect { it.file.name }
+                    println "ids: " + third.collect { it.id }
+                    println "components: " + third.collect { it.id.componentIdentifier }
+                    println "variants: " + third.collect { it.variant.attributes }
+                }
+            }
+            
+            task resolve dependsOn 'resolveFirst', 'increment', 'resolveSecond', 'resolveThird'
+        """
+
+        when:
+        run 'resolve'
+
+        then:
+        outputContains("variants: [{artifactType=firstCount}, {artifactType=firstCount}]")
+        file("build/libs1").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs1/test-1.3.jar.txt").readLines() == ["0", "1", "2", "3", "4"]
+        file("build/libs1/test2-2.3.jar.txt").readLines() == ["0", "1", "2", "3", "4"]
+
+        and:
+        outputContains("variants: [{artifactType=secondCount}, {artifactType=secondCount}]")
+        file("build/libs2").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs2/test-1.3.jar.txt").readLines() == ["1", "2", "3", "4", "5"]
+        file("build/libs2/test2-2.3.jar.txt").readLines() == ["1", "2", "3", "4", "5"]
+
+        and:
+        outputContains("variants: [{artifactType=thirdCount}, {artifactType=thirdCount}]")
+        file("build/libs3").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs3/test-1.3.jar.txt").readLines() == ["2", "3", "4", "5", "6"]
```

Comment:
I *think* you aren't reading the setup correctly. But, maybe we should chat about this.

In between registering each of the three transforms, I mutate the state of `counter` with the `increment()` method. The first transform snapshots the state of the `counter` with value `0`. The second with `1` and the third with `2`.  So, each of the two files which are transformed with the same transform are producing the same outputs. For example, both `libs3/test-1.3.jar.txt` and `libs3/test2-2.3.jar.txt` end up with lines `["2", "3", "4", "5", "6"]` because they were both registered when the counter had the value `2`.  It's important that even though the transform itself is mutating the counter's state by calling `increment()` the transform that is run for the other file, doesn't see the mutation.